File: backend/src/middleware/verify_supabase_token.ts
import { Request, Response, NextFunction } from 'express';
import { SupabaseClient, createClient } from '@supabase/supabase-js';

import { getUserSupabaseClient } from '@/lib/supabase.js';

export type SupabaseUser = { id: string; email: string; role: string };

export interface AuthRequest extends Request {
  user?: SupabaseUser;
  userClient?: SupabaseClient;
}

function parseBearerToken(authHeader: string | undefined): string {
  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    throw new Error('Missing or invalid Authorization header');
  }
  return authHeader.slice('Bearer '.length).trim();
}

function createAnonSupabaseClient() {
  const url = process.env.SUPABASE_URL?.trim() || process.env.NEXT_PUBLIC_SUPABASE_URL?.trim();
  const anonKey =
    process.env.SUPABASE_ANON_KEY?.trim() || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY?.trim();

  if (!url || !anonKey) {
    throw new Error('Missing Supabase environment variables');
  }

  return createClient(url, anonKey, {
    auth: {
      autoRefreshToken: false,
      persistSession: false,
    },
  });
}

export async function verifySupabaseToken(
  authHeader: string | undefined
): Promise<SupabaseUser> {
  const token = parseBearerToken(authHeader);
  const supabase = createAnonSupabaseClient();
  const {
    data: { user },
    error,
  } = await supabase.auth.getUser(token);

  if (error) {
    throw new Error(error.message);
  }
  if (!user) {
    throw new Error('Invalid or expired token');
  }

  const result: SupabaseUser = {
    id: user.id,
    email: user.email ?? '',
    role: user.role ?? '',
  };
  console.info(`[auth] verified user_id=${result.id}`);
  return result;
}

export async function requireAuth(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const authHeader = req.headers.authorization;
    const user = await verifySupabaseToken(authHeader);
    const authReq = req as AuthRequest;
    authReq.user = user;
    authReq.userClient = getUserSupabaseClient(parseBearerToken(authHeader));
    res.locals.user = user;
    next();
  } catch (error) {
    const message =
      error instanceof Error && error.message ? error.message : 'Authentication failed';
    res.status(401).json({ error: message });
  }
}

File: backend/src/middleware/verify_supabase_token.py
import asyncio
import logging
import os
from dataclasses import dataclass
from typing import Any

from fastapi import Header, HTTPException, Request
from supabase import AsyncClient, acreate_client

from services.config.load_env import load_root_env
from services.db.supabase_client import get_user_supabase_client

load_root_env()

logger = logging.getLogger(__name__)

# IMPORTANT: This client is ONLY for auth.get_user() token verification.
# It uses the service-role key. Never use it for data reads — it bypasses RLS.
_auth_verification_client: AsyncClient | None = None
_client_lock = asyncio.Lock()


def _read_supabase_url() -> str | None:
    return os.environ.get("SUPABASE_URL") or os.environ.get("NEXT_PUBLIC_SUPABASE_URL")


def _read_service_role_key() -> str | None:
    return os.environ.get("SUPABASE_SERVICE_ROLE_KEY")


async def _get_async_supabase_client() -> AsyncClient:
    global _auth_verification_client
    if _auth_verification_client is not None:
        return _auth_verification_client

    async with _client_lock:
        if _auth_verification_client is not None:
            return _auth_verification_client

        url = _read_supabase_url()
        service_role_key = _read_service_role_key()
        if not url or not service_role_key:
            raise HTTPException(
                status_code=401,
                detail="Missing Supabase environment variables",
            )

        _auth_verification_client = await acreate_client(url, service_role_key)
        return _auth_verification_client


@dataclass
class SupabaseUser:
    id: str
    email: str
    role: str
    user_id: str
    client: Any


async def verify_supabase_token(authorization: str | None) -> SupabaseUser:
    if authorization is None or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Missing or invalid Authorization header")

    token = authorization[len("Bearer "):].strip()
    if not token:
        raise HTTPException(status_code=401, detail="Missing or invalid Authorization header")

    client = await _get_async_supabase_client()
    try:
        response = await client.auth.get_user(token)
    except Exception as exc:
        raise HTTPException(status_code=401, detail="Invalid or expired token") from exc

    user = response.user if response else None
    if user is None:
        error_message = "Invalid or expired token"
        error_obj = getattr(response, "error", None) if response else None
        if error_obj is not None:
            error_message = getattr(error_obj, "message", error_message) or error_message
        raise HTTPException(status_code=401, detail=error_message)

    role = getattr(user, "role", "") or ""
    logger.debug("FastAPI auth verified user_id=%s", user.id)
    return SupabaseUser(
        id=str(user.id),
        email=str(getattr(user, "email", "") or ""),
        role=str(role),
        user_id=str(user.id),
        client=get_user_supabase_client(token),
    )


async def require_auth(
    request: Request,
    authorization: str = Header(None),
) -> SupabaseUser:
    user = await verify_supabase_token(authorization)
    request.state.user = user
    return user
